### ISSUE TITLE
@torch.no_grad() and with torch.no_grad() seem to have no effect

### ISSUE THREAD

**Issue by zhaomingheyuhan** [huggingface/accelerate#2118]
![1699001315387](https://github.com/huggingface/accelerate/assets/68057146/60464fa3-59f5-4541-9767-e2cd44abdebc)

If I add @torch.no_grad() or with torch.no_grad(), the error shown in the above figure will occur.
I replace 'accelerate launch --multi_gpu main.py .....' with 'python main.py', the error disappears.

**Comment by BenjaminBossan** [comment]
Where did you add the decorator/context manager? Without showing a more complete code example, it's hard for us to help. Also, what are your accelerate settings?

**Comment by zhaomingheyuhan** [comment]
**System Info**
- `Accelerate` version: 0.23.0
- Platform: Linux dsw-2408-5799c4f87f-ksxbb 3.10.0-1160.el7.x86_64
- Python version: 3.9.18
- Numpy version: 1.24.3
- PyTorch version : 1.11.0+cu113
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 512 GB
- GPU type: A100
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: fp16
	- use_cpu: False
	- debug: False
	- num_processes: 2
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
	- tpu_env: []
**Information**
 My own modified scripts
**Tasks**
 My own task or dataset (give details below)


`

class ZoeD(nn.Module):
    def __init__(self):
        super().__init__()
        conf = get_config("zoedepth", "infer")
        self.model = build_model(conf).to('cuda')  # The forward function of the model has also been added @torch.no_grad()

    #@torch.no_grad()
    def forward(self, img):
        with torch.no_grad():
            self.model.eval()
            dep_fea = self.model.infer_pil(img)
        return dep_fea


class PointCloudProjectionModel(ModelMixin):
    def __init__(
        self, ...)
        ....
        self.zoefea = ZoeD()

    def get_input_with_conditioning(...):
        zoe = self.zoefea(image_rgb)  
    
    def forward(self, batch: FrameData, mode: str = 'train', **kwargs):
        """ The forward method may be defined differently for different models. """
        raise NotImplementedError()

`

**Error:** 
![1699083885563](https://github.com/huggingface/accelerate/assets/68057146/870db21a-f836-482a-86de-d8e802a79955)


I replace 'accelerate launch --multi_gpu main.py .....' with 'python main.py', the error disappears.

**Comment by zhaomingheyuhan** [comment]
> Where did you add the decorator/context manager? Without showing a more complete code example, it's hard for us to help. Also, what are your accelerate settings?

**System Info**
- `Accelerate` version: 0.23.0
- Platform: Linux dsw-2408-5799c4f87f-ksxbb 3.10.0-1160.el7.x86_64
- Python version: 3.9.18
- Numpy version: 1.24.3
- PyTorch version : 1.11.0+cu113
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 512 GB
- GPU type: A100
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: fp16
	- use_cpu: False
	- debug: False
	- num_processes: 2
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
	- tpu_env: []
	-

**Information**
 My own modified scripts
**Tasks**
 My own task or dataset (give details below)


```
class ZoeD(nn.Module):
    def __init__(self):
        super().__init__()
        conf = get_config("zoedepth", "infer")
        self.model = build_model(conf).to('cuda')  # The forward function of the model has also been added @torch.no_grad()

    #@torch.no_grad()
    def forward(self, img):
        with torch.no_grad():
            self.model.eval()
            dep_fea = self.model.infer_pil(img)
        return dep_fea


class PointCloudProjectionModel(ModelMixin):
    def __init__(
        self, ...)
        ....
        self.zoefea = ZoeD()

    def get_input_with_conditioning(...):
        zoe = self.zoefea(image_rgb)  
    
    def forward(self, batch: FrameData, mode: str = 'train', **kwargs):
        """ The forward method may be defined differently for different models. """
        raise NotImplementedError()
```


**Error:** 
![1699083885563](https://github.com/huggingface/accelerate/assets/68057146/870db21a-f836-482a-86de-d8e802a79955)


I replace 'accelerate launch --multi_gpu main.py .....' with 'python main.py', the error disappears.

**Comment by zhaomingheyuhan** [comment]
> Where did you add the decorator/context manager? Without showing a more complete code example, it's hard for us to help. Also, what are your accelerate settings?

**System Info**
- `Accelerate` version: 0.23.0
- Platform: Linux dsw-2408-5799c4f87f-ksxbb 3.10.0-1160.el7.x86_64
- Python version: 3.9.18
- Numpy version: 1.24.3
- PyTorch version : 1.11.0+cu113
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 512 GB
- GPU type: A100
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: fp16
	- use_cpu: False
	- debug: False
	- num_processes: 2
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
	- tpu_env: []
	-

**Information**
 My own modified scripts
**Tasks**
 My own task or dataset (give details below)


```
class ZoeD(nn.Module):
    def __init__(self):
        super().__init__()
        conf = get_config("zoedepth", "infer")
        self.model = build_model(conf).to('cuda')  # The forward function of the model has also been added @torch.no_grad()

    #@torch.no_grad()
    def forward(self, img):
        with torch.no_grad():
            self.model.eval()
            dep_fea = self.model.infer_pil(img)
        return dep_fea


class PointCloudProjectionModel(ModelMixin):
    def __init__(
        self, ...)
        ....
        self.zoefea = ZoeD()

    def get_input_with_conditioning(...):
        zoe = self.zoefea(image_rgb)  
    
    def forward(self, batch: FrameData, mode: str = 'train', **kwargs):
        """ The forward method may be defined differently for different models. """
        raise NotImplementedError()
```


**Error:** 
![1699083885563](https://github.com/huggingface/accelerate/assets/68057146/870db21a-f836-482a-86de-d8e802a79955)


I replace 'accelerate launch --multi_gpu main.py .....' with 'python main.py', the error disappears.

**Comment by BenjaminBossan** [comment]
It seems that PyTorch believes that this (sub-)module should participate in the loss calculation, despite the precautions that you take. Maybe after the line `self.zoefea = ZoeD()`, you could try calling `self.zoefea.requires_grad_(False)` and check if that helps. It's also important to ensure that nowhere else in the code, gradients for this module are enabled.

**Comment by zhaomingheyuhan** [comment]
> It seems that PyTorch believes that this (sub-)module should participate in the loss calculation, despite the precautions that you take. Maybe after the line `self.zoefea = ZoeD()`, you could try calling `self.zoefea.requires_grad_(False)` and check if that helps. It's also important to ensure that nowhere else in the code, gradients for this module are enabled.

It works. Thank you very much!
